impeller/entity/save_layer_utils.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/entity/save_layer_utils.h"
+
+namespace impeller {
+
+std::optional<Rect> ComputeSaveLayerCoverage(
+    const Rect& content_coverage,
+    const Matrix& effect_transform,
+    const Rect& coverage_limit,
+    const std::shared_ptr<FilterContents>& image_filter,
+    bool flood_output_coverage,
+    bool flood_input_coverage) {
+  Rect coverage = content_coverage;
+  // There are two conditions that currently trigger flood_input_coverage, the
+  // first is the presence of a backdrop filter and the second is the presence
+  // of a color filter that effects transparent black.
+  //
+  // Backdrop filters apply before the saveLayer is restored. The presence of
+  // a backdrop filter causes the content coverage of the saveLayer to be
+  // unbounded.
+  //
+  // If there is a color filter that needs to flood its output. The color filter
+  // is applied before any image filters, so this floods input coverage and not
+  // the output coverage. Technically, we only need to flood the output of the
+  // color filter and could allocate a render target sized just to the content,
+  // but we don't currenty have the means to do so. Flooding the coverage is a
+  // non-optimal but technically correct way to render this.
+  if (flood_input_coverage) {
+    coverage = Rect::MakeMaximum();
+  }
+
+  // If flood_output_coverage is true, then the restore is applied with a
+  // destructive blend mode that requires flooding to the coverage limit.
+  // Technically we could only allocated a render target as big as the input
+  // coverage and then use a decal sampling mode to perform the flood. Returning
+  // the coverage limit is a correct but non optimal means of ensuring correct
+  // rendering.
+  if (flood_output_coverage) {
+    return coverage_limit;
+  }
+
+  // The content coverage must be scaled by any image filters present on the
+  // saveLayer paint. For example, if a saveLayer has a coverage limit of
+  // 100x100, but it has a Matrix image filter that scales by one half, the
+  // actual coverage limit is 200x200.
+  if (image_filter) {
+    std::optional<Rect> source_coverage_limit =
+        image_filter->GetSourceCoverage(effect_transform, coverage_limit);
+    if (!source_coverage_limit.has_value()) {
+      // No intersection with parent coverage limit.
+      return std::nullopt;
+    }
```

Comment:
I'm trying to wrap my head around this. Why aren't we handing in content_coverage? That might be the same as coverage_limit if there was flooding, but it could be less.

Reply:
Not sure what you mean exactly. A case where this would fire is, you an IF that translates a SL entirely offscreen, or somethng like that.

Reply:
I moved the comment around a bit so it should be clearer.